internal/femto/femto.go

```diff
@@ -3,124 +3,145 @@ package femto
-	f.mux.HandleFunc(pattern, func(w http.ResponseWriter, r *http.Request) {
-		doPost(f, w, r, handle)
+type PostFunc[T any] func(T, *http.Request, *slog.Logger) (*HTTPResponseContent[types.Unit], error)
+type GetFunc[T any] func(*http.Request, *slog.Logger) (*HTTPResponseContent[T], error)
+
+func (femto *Femto) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	femto.mux.ServeHTTP(w, r)
+}
+
+func OnPost[T any](femto *Femto, pattern string, handle PostFunc[T]) {
+	femto.mux.HandleFunc(pattern, func(writer http.ResponseWriter, request *http.Request) {
+		doPost(femto, writer, request, handle)
 	})
 }
 
-func OnGet[T any](f *Femto, pattern string, handle GetFunc[T]) {
-	f.mux.HandleFunc(pattern, func(w http.ResponseWriter, r *http.Request) {
-		doGet(f, w, r, handle)
+func OnGet[T any](femto *Femto, pattern string, handle GetFunc[T]) {
+	femto.mux.HandleFunc(pattern, func(writer http.ResponseWriter, request *http.Request) {
+		doGet(femto, writer, request, handle)
 	})
 }
 
-func doGet[T any](f *Femto, w http.ResponseWriter, r *http.Request, handle GetFunc[T]) {
-	reqNo := f.nextReqNo()
-	log := f.logger().With(slog.Uint64("req_no", reqNo))
+func generateErrorLogger(log *slog.Logger, writer http.ResponseWriter) func(ctx string, status int, e error) {
+	return func(ctx string, status int, e error) {
+		log.Error(ctx, "err", e)
+		http.Error(writer, e.Error(), status)
+	}
+}
+
+func doGet[T any](femto *Femto, writer http.ResponseWriter, request *http.Request, handle GetFunc[T]) {
+	reqNo := femto.nextReqNo()
+	log := femto.logger().With(slog.Uint64("req_no", reqNo))
 
-	log.Info("New Request", "method", r.Method, "url", r.URL, "from", r.RemoteAddr)
+	log.Info("New Request", "method", request.Method, "url", request.URL, "from", request.RemoteAddr)
 
-	if r.Method != http.MethodGet {
+	if request.Method != http.MethodGet {
 		log.Info("Wanted GET, returned 405")
-		http.Error(w, "Expected GET", http.StatusMethodNotAllowed)
+		http.Error(writer, "Expected GET", http.StatusMethodNotAllowed)
 		return
 	}
 
-	ise := func(ctx string, e error) {
-		log.Error(ctx, "err", e)
-		http.Error(w, e.Error(), http.StatusInternalServerError)
+	data, err := handle(request, log)
+
+	if data == nil {
+		data = &HTTPResponseContent[T]{}
 	}
 
-	data, err := handle(r, log)
+	if data.Status == 0 {
+		data.Status = http.StatusInternalServerError
+	}
+
+	ise := generateErrorLogger(log, writer)
+
 	if err != nil {
-		// Handle authentication related errors
-		if errors.Is(err, NotAuthenticatedError) {
-			log.Info("Invalid attempt at authentication")
-			http.Error(w, "Not authenticated", http.StatusUnauthorized)
-			return
-		}
-		ise("application error", err)
+		ise("There was an error in the handler", data.Status, err)
 		return
 	}
 
-	jsonb, err := json.Marshal(data)
+	jsonb, err := json.Marshal(data.Body)
 	if err != nil {
-		ise("marshaling to json failed", err)
+		ise("There was an error in trying to parse the handler's response into JSON", data.Status, err)
```

Comment:
But we're not parsing to JSON, we're marshaling. Please keep the reference to marshaling, even if you feel the need to change the errors.

Reply:
This is a user-facing error via `ise`. It's Go-specific error messaging to note marshalling. To the user, this is a web API and so we should prioritise idiomatic web-oriented terminology rather than that of Go.

Reply:
Sure, but we're still not parsing. Parsing is string->structured, and this is structured->string. If you want a less go specific term, that's fine, as long as it's right.